src/v/archival/scrubber.cc

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "archival/scrubber.h"
+
+#include "archival/logger.h"
+#include "cloud_storage/partition_manifest.h"
+#include "cloud_storage/remote.h"
+#include "cloud_storage/remote_partition.h"
+#include "cloud_storage/topic_manifest.h"
+#include "cloud_storage/tx_range_manifest.h"
+#include "cluster/members_table.h"
+#include "cluster/topic_table.h"
+#include "cluster/topics_frontend.h"
+#include "config/configuration.h"
+#include "config/node_config.h"
+#include "hashing/xx.h"
+#include "vlog.h"
+
+namespace archival {
+
+using cloud_storage::download_result;
+using cloud_storage::upload_result;
+
+scrubber::scrubber(
+  cloud_storage::remote& r,
+  cluster::topic_table& tt,
+  ss::sharded<cluster::topics_frontend>& tf,
+  ss::sharded<cluster::members_table>& mt)
+  : _api(r)
+  , _topic_table(tt)
+  , _topics_frontend(tf)
+  , _members_table(mt) {}
+
+ss::future<scrubber::purge_result> scrubber::purge_partition(
+  const cloud_storage_clients::bucket_name& bucket,
+  model::ntp ntp,
+  model::initial_revision_id remote_revision,
+  retry_chain_node& parent_rtc) {
+    retry_chain_node manifest_rtc(
+      ss::lowres_clock::now() + 5s, 1s, &parent_rtc);
+
+    purge_result result{.status = purge_status::success, .ops = 0};
+
+    // TODO: tip off remote() to not retry on SlowDown responses: if we hit
+    // one during housekeeping we should drop out.  Maybe retry_chain_node
+    // could have a "drop out on slowdown" flag?
+
+    cloud_storage::partition_manifest manifest(ntp, remote_revision);
+    auto manifest_path = manifest.get_manifest_path();
+    auto manifest_get_result = co_await _api.maybe_download_manifest(
+      bucket, manifest_path, manifest, manifest_rtc);
+
+    if (manifest_get_result == download_result::notfound) {
+        vlog(
+          archival_log.debug,
+          "Partition manifest get {} not found",
+          manifest_path);
+        result.status = purge_status::permanent_failure;
+        co_return result;
+    } else if (manifest_get_result != download_result::success) {
+        vlog(
+          archival_log.debug,
+          "Partition manifest get {} failed: {}",
+          manifest_path,
+          manifest_get_result);
+        result.status = purge_status::retryable_failure;
+        co_return result;
+    }
+
+    auto partition_r = co_await cloud_storage::remote_partition::erase(
+      _api, bucket, std::move(manifest), _as);
```

Comment:
I'm treating the housekeeping ops counter as something more like an abstract "fuel" than a literal number of ops, although it's a fair point that erase() can be rather expensive  if a partition really has many thousands of segments.

This is going to be a sticking point in general, because we don't know without going all the way down to the S3/ABS code how many underlying requests were actually required to delete some objects :-/

We may need some discussion about this on the following PR as well, in terms of what counts as an op, but for now I have added a couple lines here in an extra commit, which account for the ops in remote_partition::erase as the number of segs / 1000, which is roughly true in the AWS S3 case.  The actual amount of quota we get is pretty small under normal circumstances, so this will make scrubbing topics with lots of partitions take many iterations (which is probably fine)

Reply:
I think this is reasonable. It might make us get throttled sooner on GCP, but we can refine this in a follow-up.